### ISSUE TITLE
Session Persistance

### ISSUE THREAD

**Issue by martin9700** [IAG-NZ/jenkins#101]
Having to provide the URL, Credential and Crumb information for every command is a bit of a pain and not terribly user friendly.  

I'd like to create 2 functions to address this:
Set-JenkinsConfig: which would drop a simple text file in the users profile with the URL in it and whether or not a Crumb is needed.
New-JenkinsSession: this would create a variable in the global scope containing the user's credential object and crumb (if needed) in it. 

There would be two ways that this session information could be used.  

1. Create a helper script that would run at the beginning of every function to check for that variable. If it doesn't exist run New-JenkinsSession.  If it does exist simply return.  This would mean the URI, Credential and Crumb parameters would no longer be needed.  For backwards compatibility--at least for awhile--the big 3 parameters would stay but simply be passed to the helper script to create the variable (or ignored completely).

2. Add a parameter to each function called Session that could be used instead of the big three.  A parameterset could be used so it's either the session variable or the big 3.

I'd like to hear from the Architects and community on appetite for this?  Option 1 is probably more disruptive, but from experience makes using the module a far more interactive possibility.  Option 2 is less disruptive but would still require the user to remember the variable name--and maybe instead of a global it's just a captured at runtime and the user is required to do that work--and enter it each time a function is used.

**Comment by PlagueHO** [comment]
Hi @martin9700 - some good ideas there. Let me have a think about it and leave it open for a bit for community comment.

I'm tending towards option 2 in the short term with the possibility of looking at an option 3 using a model closer to how Kubectl or Azure PowerShell uses.

In the meantime, you could fairly easily use the $PSDefaultParameterValues settings to default parameters you send to Jenkins cmdlets with:

```powershell
$PSDefaultParameterValues = @{"*-Jenkins*:Credential"=$credential}
$PSDefaultParameterValues = @{"*-Jenkins*:Crumb"=$crumb}
```

**Comment by martin9700** [comment]
@PlagueHO Good idea! Had not thought of that.  And once the URI is set with Set-JenkinsConfig the same technique could be used for URI, it would just be set at module load (assuming the configuration file was present).
